scrapy/core/scraper.py

```diff
@@ -96,15 +182,19 @@ def finish_response(
     ) -> None:
         self.active.remove(request)
         if isinstance(result, Response):
-            self.active_size -= max(len(result.body), self.MIN_RESPONSE_SIZE)
+            self._active_size -= max(len(result.body), self._MIN_RESPONSE_SIZE)
         else:
-            self.active_size -= self.MIN_RESPONSE_SIZE
+            self._active_size -= self._MIN_RESPONSE_SIZE
 
     def is_idle(self) -> bool:
         return not (self.queue or self.active)
 
     def needs_backout(self) -> bool:
-        return self.active_size > self.max_active_size
+        warn(
+            "scrapy.core.scraper.Slot.needs_backout is deprecated.",
+            ScrapyDeprecationWarning,
+        )
+        return self._active_size > self._max_active_size
```

Comment:
In a context of https://github.com/scrapy/scrapy/issues/482
Other important observation from log output from https://github.com/scrapy/scrapy/issues/6206#issuecomment-2180631979 is that response is active(it's size in `scraper.slot.active_size`) until all it's items passed pipelines (including awaiting for additional req/responses from media/files/images pipeline to be finished).
And by the moment when that items reach FeedExport (S3) - it's response is not counted in `scraper.slot.active_size` (and next request can be already sent without waiting to upload remaining items to S3 - which I assume required to be delayed here).

FeedExport(S3) are completely independent from scraper slot active size limits


In context of S3. It means that `scraper.slot._needs_backout`
https://github.com/scrapy/scrapy/blob/d8ecd28c5557e27f42e00bd1223b457468ea2ea7/scrapy/core/scraper.py#L106C1-L107C55
also needs to check that remaining items are exported (send to S3) 

and to be looked like 

```
    def needs_backout(self) -> bool:
        self.crawler.stats.get(')
        s3_stats_condition = #### check numbers of crawler.stats - feedexport(uploaded items) and crawler.stats (items)
        ## if item produced bud didn't uploaded to s3 yet - this numbers must be different
        return self.active_size > self.max_active_size or s3_stats_condition
```
Where `scraper.slot` class also need to have access to `crawler.stats`

Reply:
[The response is available in the context that calls item pipelines](https://github.com/scrapy/scrapy/blob/d8ecd28c5557e27f42e00bd1223b457468ea2ea7/scrapy/core/scraper.py#L375), so I think we are covered here (but tests should be written to make sure that does not change in the future accidentally).

Reply:
My assumption is that root-cause of https://github.com/scrapy/scrapy/issues/482 is that application simultaneously (trying to) do:
1. scheduling new requests and processing responses (with RAM allocations to responses)
2. sending scraped items to S3 (with related RAM allocations)
Because associated response can be not active (in terms of scraper) when it's items reach S3 exporter.

If this theory is true then application needs to make only one of this at specific moment on time (which means that scraper needs to count work progress of S3 feed.. in it's backout condition)

Reply:
That’s correct. That’s the reason I say this is part 1 of a 2-part fix, where the second part would be to ensure the response objects are not garbage-collected before S3 finishes uploading. I’m moreover thinking that we could make that upload use response.body instead of a copy [as it does now](https://github.com/scrapy/scrapy/blob/d8ecd28c5557e27f42e00bd1223b457468ea2ea7/scrapy/pipelines/files.py#L676), making the uploads more memory efficient.

Reply:
> sending scraped items to S3

This is about the media pipeline, i.e. the problem is sending *responses* to S3, not *items*. But it is true that this PR does not fix that. That’s the reason I say this is part 1 of a 2-part fix, where the second part would be to ensure the response objects are not garbage-collected before S3 finishes uploading. I’m moreover thinking that we could make that upload use response.body instead of a copy [as it does now](https://github.com/scrapy/scrapy/blob/d8ecd28c5557e27f42e00bd1223b457468ea2ea7/scrapy/pipelines/files.py#L676), making the uploads more memory efficient.